contracts/asset-proxy/contracts/src/ERC1155Proxy.sol

```diff
@@ -18,349 +18,71 @@
-                    mstore(96, 0)
-                    revert(0, 100)
-                }
-
-                // Copy `data` from `assetData` (Table #2) to memory (Table #3)
-                calldatacopy(
-                    dataAreaEndOffset,
-                    dataOffset,
-                    add(dataLengthInBytes, 32)
-                )
-
-                // Update the end of data offset to be word-aligned
-                let dataLengthInWords := div(add(dataLengthInBytes, 31), 32)
-                let dataLengthInBytesWordAligned := mul(dataLengthInWords, 32)
-                dataAreaEndOffset := add(dataAreaEndOffset, add(dataLengthInBytesWordAligned, 32))
-
-                ////////// STEP 3/3 - Execute Transfer //////////
-                // Load the address of the destination erc1155 contract from asset data (Table #2)
-                // +32 bytes for assetData Length
-                // +4 bytes for assetProxyId
-                let assetAddress := and(
-                    calldataload(add(assetDataOffset, 36)),
-                    0x000000000000000000000000ffffffffffffffffffffffffffffffffffffffff
-                )
-
-                // Call into the destination erc1155 contract using as calldata Table #3 (constructed in-memory above)
-                let success := call(
-                    gas,                                    // forward all gas
-                    assetAddress,                           // call address of erc1155 asset
-                    0,                                      // don't send any ETH
-                    0,                                      // pointer to start of input
-                    dataAreaEndOffset,                      // length of input is the end of the Data Area (Table #3)
-                    0,                                      // write output over memory that won't be reused
-                    0                                       // don't copy output to memory
-                )
-
-                // Revert with reason given by AssetProxy if `transferFrom` call failed
-                if iszero(success) {
-                    returndatacopy(
-                        0,                // copy to memory at 0
-                        0,                // copy from return data at 0
-                        returndatasize()  // copy all return data
-                    )
-                    revert(0, returndatasize())
-                }
-
-                // Return if call was successful
-                return(0, 0)
-            }
-
-            // Revert if undefined function is called
-            revert(0, 0)
+        // Decode params from `assetData`
+        // solhint-disable indent
+        (
+            address erc1155TokenAddress,
+            uint256[] memory ids,
+            uint256[] memory values,
+            bytes memory data
+        ) = abi.decode(
+            assetData.sliceDestructive(4, assetData.length),
```

Comment:
I'm surprised this works with `assetData` being `calldata` and not `memory`. Is it cloning `assetData` to memory at the time of this call?

Reply:
I believe calldata gets copied to memory when you pass it into any internal function (`sliceDestructive` in this case). That might not be implemented for things that require ABIv2 encoding, though (or maybe just structs?).